services/wallet/service.go

```diff
@@ -689,6 +719,87 @@ func (service *Service) LinkUpholdWallet(ctx context.Context, wallet uphold.Wall
 	return country, nil
 }
 
+func (service *Service) LinkSolanaAddress(ctx context.Context, paymentID uuid.UUID, req linkSolanaAddrRequest) error {
+	if err := isWalletWhitelisted(ctx, service.Datastore.RawDB(), service.allowListRepo, paymentID); err != nil {
+		return err
+	}
+
+	ctx, txn, rollback, commit, err := getTx(ctx, service.Datastore)
+	if err != nil {
+		return err
+	}
+	defer rollback()
+
+	chl, err := service.chlRepo.Get(ctx, txn, paymentID.String())
+	if err != nil {
+		return err
+	}
+
+	if err := chl.IsValid(time.Now()); err != nil {
+		return err
```

Comment:
Should we delete the challenge here if this errors out too?

Reply:
We could do this although it will add a bit more complexity to which error we return. There is going to be a separate job which removes all expired/failed challenges, I will add this to the [issue](https://github.com/brave-intl/payment-services/issues/185)